File: backend/src/AgentTeam/Services/StreamContext.php
<?php

declare(strict_types=1);

namespace AgentTeam\Services;

/**
 * Stream Context
 *
 * Holds the SSE callback and provides methods for emitting agent activity events.
 * This context is passed through the delegation chain to enable real-time updates.
 */
class StreamContext
{
    /** @var callable|null */
    private $onEvent = null;
    private int $userId = 0;
    private ?int $rootExecutionId = null;

    public function __construct(?callable $onEvent = null, int $userId = 0)
    {
        $this->onEvent = $onEvent;
        $this->userId = $userId;
    }

    /**
     * Set the event callback
     */
    public function setEventCallback(callable $callback): self
    {
        $this->onEvent = $callback;
        return $this;
    }

    /**
     * Set the root execution ID (the top-level manager execution)
     */
    public function setRootExecutionId(int $executionId): self
    {
        $this->rootExecutionId = $executionId;
        return $this;
    }

    /**
     * Get the root execution ID
     */
    public function getRootExecutionId(): ?int
    {
        return $this->rootExecutionId;
    }

    /**
     * Get user ID
     */
    public function getUserId(): int
    {
        return $this->userId;
    }

    /**
     * Check if streaming is enabled
     */
    public function isStreaming(): bool
    {
        return $this->onEvent !== null;
    }

    /**
     * Emit an agent_start event
     */
    public function emitAgentStart(
        int $agentId,
        string $agentName,
        string $agentType,
        ?int $parentAgentId = null,
        ?int $executionId = null
    ): void {
        $this->emit([
            'type' => 'agent_start',
            'agent_id' => $agentId,
            'agent_name' => $agentName,
            'agent_type' => $agentType,
            'parent_agent_id' => $parentAgentId,
            'execution_id' => $executionId,
            'timestamp' => microtime(true),
        ]);
    }

    /**
     * Emit an agent_delegate event
     */
    public function emitAgentDelegate(
        int $fromAgentId,
        string $fromAgentName,
        int $toAgentId,
        string $toAgentName,
        string $task
    ): void {
        $this->emit([
            'type' => 'agent_delegate',
            'from_agent_id' => $fromAgentId,
            'from_agent_name' => $fromAgentName,
            'to_agent_id' => $toAgentId,
            'to_agent_name' => $toAgentName,
            'task' => mb_substr($task, 0, 200) . (mb_strlen($task) > 200 ? '...' : ''),
            'timestamp' => microtime(true),
        ]);
    }

    /**
     * Emit an agent_complete event
     */
    public function emitAgentComplete(
        int $agentId,
        string $agentName,
        string $agentType,
        bool $success = true,
        ?string $error = null,
        ?int $executionId = null
    ): void {
        $this->emit([
            'type' => 'agent_complete',
            'agent_id' => $agentId,
            'agent_name' => $agentName,
            'agent_type' => $agentType,
            'success' => $success,
            'error' => $error,
            'execution_id' => $executionId,
            'timestamp' => microtime(true),
        ]);
    }

    /**
     * Emit an agent_thinking event (for progress indication)
     */
    public function emitAgentThinking(
        int $agentId,
        string $agentName,
        string $status = 'thinking'
    ): void {
        $this->emit([
            'type' => 'agent_thinking',
            'agent_id' => $agentId,
            'agent_name' => $agentName,
            'status' => $status,
            'timestamp' => microtime(true),
        ]);
    }

    /**
     * Emit a text chunk (for streaming responses)
     */
    public function emitChunk(string $text, int $agentId, string $agentName): void
    {
        $this->emit([
            'type' => 'chunk',
            'text' => $text,
            'agent_id' => $agentId,
            'agent_name' => $agentName,
        ]);
    }

    /**
     * Emit an error event
     */
    public function emitError(string $error, ?int $agentId = null): void
    {
        $this->emit([
            'type' => 'error',
            'error' => $error,
            'agent_id' => $agentId,
            'timestamp' => microtime(true),
        ]);
    }

    /**
     * Emit a generic event
     */
    public function emit(array $data): void
    {
        if ($this->onEvent !== null) {
            error_log("[StreamContext] Emitting event: " . ($data['type'] ?? 'unknown') . " - " . json_encode($data));
            ($this->onEvent)($data);
        } else {
            error_log("[StreamContext] WARNING: No callback set, cannot emit: " . ($data['type'] ?? 'unknown'));
        }
    }
}

File: backend_python/app/agent_team/services/stream_context.py
"""Port of backend/src/AgentTeam/Services/StreamContext.php.

Holds the SSE callback and provides methods for emitting agent activity events.
This context is passed through the delegation chain to enable real-time updates.
"""
from __future__ import annotations

import time
from typing import Callable, Optional

from app.support.logger import error_log
from app.support.phpcompat import mb_substr
from app.support.phpjson import php_json_encode


class StreamContext:
    def __init__(self, on_event: Optional[Callable[[dict], None]] = None, user_id: int = 0):
        self.onEvent = on_event
        self.userId = user_id
        self.rootExecutionId: int | None = None

    def setEventCallback(self, callback: Callable[[dict], None]) -> 'StreamContext':
        self.onEvent = callback
        return self

    def setRootExecutionId(self, execution_id: int) -> 'StreamContext':
        self.rootExecutionId = execution_id
        return self

    def getRootExecutionId(self) -> int | None:
        return self.rootExecutionId

    def getUserId(self) -> int:
        return self.userId

    def isStreaming(self) -> bool:
        return self.onEvent is not None

    def emitAgentStart(
        self,
        agent_id: int,
        agent_name: str,
        agent_type: str,
        parent_agent_id: int | None = None,
        execution_id: int | None = None,
    ) -> None:
        self.emit({
            'type': 'agent_start',
            'agent_id': agent_id,
            'agent_name': agent_name,
            'agent_type': agent_type,
            'parent_agent_id': parent_agent_id,
            'execution_id': execution_id,
            'timestamp': time.time(),
        })

    def emitAgentDelegate(
        self,
        from_agent_id: int,
        from_agent_name: str,
        to_agent_id: int,
        to_agent_name: str,
        task: str,
    ) -> None:
        self.emit({
            'type': 'agent_delegate',
            'from_agent_id': from_agent_id,
            'from_agent_name': from_agent_name,
            'to_agent_id': to_agent_id,
            'to_agent_name': to_agent_name,
            'task': mb_substr(task, 0, 200) + ('...' if len(task) > 200 else ''),
            'timestamp': time.time(),
        })

    def emitAgentComplete(
        self,
        agent_id: int,
        agent_name: str,
        agent_type: str,
        success: bool = True,
        error: str | None = None,
        execution_id: int | None = None,
    ) -> None:
        self.emit({
            'type': 'agent_complete',
            'agent_id': agent_id,
            'agent_name': agent_name,
            'agent_type': agent_type,
            'success': success,
            'error': error,
            'execution_id': execution_id,
            'timestamp': time.time(),
        })

    def emitAgentThinking(
        self,
        agent_id: int,
        agent_name: str,
        status: str = 'thinking',
    ) -> None:
        self.emit({
            'type': 'agent_thinking',
            'agent_id': agent_id,
            'agent_name': agent_name,
            'status': status,
            'timestamp': time.time(),
        })

    def emitChunk(self, text: str, agent_id: int, agent_name: str) -> None:
        self.emit({
            'type': 'chunk',
            'text': text,
            'agent_id': agent_id,
            'agent_name': agent_name,
        })

    def emitError(self, error: str, agent_id: int | None = None) -> None:
        self.emit({
            'type': 'error',
            'error': error,
            'agent_id': agent_id,
            'timestamp': time.time(),
        })

    def emit(self, data: dict) -> None:
        if self.onEvent is not None:
            event_type = data.get('type') if data.get('type') is not None else 'unknown'
            error_log(f'[StreamContext] Emitting event: {event_type} - {php_json_encode(data)}')
            self.onEvent(data)
        else:
            event_type = data.get('type') if data.get('type') is not None else 'unknown'
            error_log(f'[StreamContext] WARNING: No callback set, cannot emit: {event_type}')
